actions/tamper_action.go

```diff
@@ -86,60 +89,637 @@ func ParseTamperAction(s *scanner.Scanner) (Action, error) {
+		tcp.Window = uint16(valueGen.uint(16))
+	case TCPFieldUrgent:
+		tcp.Urgent = uint16(valueGen.uint(16))
+	case TCPFieldFlags:
+		setTCPFlags(tcp, uint16(valueGen.uint(16)))
+	case TCPLoad:
+		tcp.Payload = valueGen.bytes(1480)
+	default:
+		// find option in TCP header
+		var opt *layers.TCPOption
+		for i, o := range tcp.Options {
+			if field == TCPField(o.OptionType) {
+				opt = &tcp.Options[i]
+				break
+			}
+		}
+
+		// create option if it doesn't exist and move options-eol to the end of the list
+		if opt == nil {
+			tcp.Options = append(tcp.Options, layers.TCPOption{
+				OptionType: layers.TCPOptionKind(field),
+			})
+
+			ol := len(tcp.Options)
+			tcp.Options[ol-2], tcp.Options[ol-1] = tcp.Options[ol-1], tcp.Options[ol-2]
+			opt = &tcp.Options[ol-2]
+		}
+
+		opt.OptionData = valueGen.bytes(tcpOptionLengths[field])
+		if field == TCPOptionEol || field == TCPOptionNop {
+			opt.OptionLength = 1
 		} else {
-			return nil, fmt.Errorf("invalid action for tamper rule: %w", err)
+			opt.OptionLength = uint8(tcpOptionLengths[field]) + 2
+		}
+
+	}
+
+	// let gopacket handle converting the modified TCP headers into []byte for us since we changed the struct fields
+	// instead of the underlying []byte directly. SerializeTo doesn't write the changes to the raw packet
+	// so we have to copy the formatted bytes back into the packet header.
+	sb := gopacket.NewSerializeBuffer()
+	tcp.SerializeTo(sb, gopacket.SerializeOptions{})
+	tcp.Contents = make([]byte, len(sb.Bytes()))
+	copy(tcp.Contents, sb.Bytes())
+}
+
+// tcpFlagsToUint32 converts a string of TCP flags to a uint32 bitmap.
+func tcpFlagsToUint32(flags string) uint32 {
+	flags = strings.ToLower(flags)
+	var f uint32
+	for _, c := range flags {
+		switch c {
+		case 'f':
+			f |= 0x0001
+		case 's':
+			f |= 0x0002
+		case 'r':
+			f |= 0x0004
+		case 'p':
+			f |= 0x0008
+		case 'a':
+			f |= 0x0010
+		case 'u':
+			f |= 0x0020
+		case 'e':
+			f |= 0x0040
+		case 'c':
+			f |= 0x0080
+		case 'n':
+			f |= 0x0100
+		}
+	}
+	return f
+}
+
+// setTCPFlags sets the tcp struct fields using flags as a bitmap, does not modify the raw packet bytes.
+func setTCPFlags(tcp *layers.TCP, flags uint16) {
+	tcp.FIN = flags&0x0001 != 0
+	tcp.SYN = flags&0x0002 != 0
+	tcp.RST = flags&0x0004 != 0
+	tcp.PSH = flags&0x0008 != 0
+	tcp.ACK = flags&0x0010 != 0
+	tcp.URG = flags&0x0020 != 0
+	tcp.ECE = flags&0x0040 != 0
+	tcp.CWR = flags&0x0080 != 0
+	tcp.NS = flags&0x0100 != 0
+}
+
+// updateTCPDataOffAndChksum updates the TCP data offset and checksum fields on the TCP struct and in the raw packet bytes.
+func updateTCPDataOffAndChksum(tcp *layers.TCP) {
+	// update data offset
+	headerLen := len(tcp.Contents)
+	tcp.DataOffset = uint8(headerLen / 4)
+	tcp.Contents[12] = tcp.DataOffset << 4
+
+	// update checksum.
+	// the ComputeChecksum method requires the checksum bytes in the raw packet to be zeroed out.
+	tcp.Contents[16] = 0
+	tcp.Contents[17] = 0
+	chksum, _ := tcp.ComputeChecksum()
```

Comment:
I used `gopacket` to calculate checksum since I'm already using it to easily alter the packet. I figured since the packet is being passed in as a `gopacket.packet`, might as well stick to using it. I also figured it would make the code more readable and self-documenting. 

`SerializeTo()` does take an option to let it calculate the checksum but it only passed in the serialized TCP headers and not the payload to the non-exported `calculateChecksum()` function. Whereas, calling `ComputeChecksum()` directly passed both the headers and payload. `SerializeTo()` also doesn't serialize the payload to the buffer.

Reply:
Ah, gotcha. Yeah, that's cool then. With the level of effort we're willing to give to this right now, let's just ignore my comment above.